Don't silently persist `BUNDLE_WITH` and `BUNDLE_WITHOUT` envs locally

deivid-rodriguez opened the pull request:
## What was the end-user or developer problem that led to this PR?

`BUNDLE_WITH` and `BUNDLE_WITHOUT` environment variables, are silently persisted to local configuration.

## What is your fix for the problem, implemented in this PR?

Fix is to refactor `bundle install` option handling to not do this.

This PR reintroduces #3666, which was reverted by https://github.com/rubygems/rubygems/pull/4154, because it should no longer have any unexpected consequences, since configuration is now local by default if inside an application.

Fixes #3708.
Fixes #4835.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

nsmmrs commented:
For anyone else who might be scratching their heads today (I hope it's OK to comment here):

We have a custom environment, `staging`, running on Heroku, and this bugfix resulted in crashing on `bundle exec` with "Could not find [all of the development/test gems] in any of the sources".

The reason was that our image is built with `BUNDLE_WITHOUT` being set via `ARG`, which is not persisted to the final image. So our image only worked in `staging` because of the Bundler bug which was silently persisting that configuration.

Luckily, fixing this was as easy as running:
```sh
heroku config:set BUNDLE_WITHOUT=development:test --remote staging
```

deivid-rodriguez commented:
Oh, that's right! First time I fixed this, people reported something similar, but in that case they were configuring `bundle config set without development:test` (which was unintentionally global), and then thanks to this bug, that global configuration was getting copied locally and making things work. Since I also changed `bundle config` to be local now, I didn't expect that issue to be hit again, but it does break your case.

I'm sorry about it, hopefully not too many people is bitten by it, and hopefully your fix will be useful to them!

nsmmrs commented:
@deivid-rodriguez No problem! This was definitely a problem of our own creation. 😅 Thanks for the bugfix!

gugaiz commented:
After this change, when deploying to our staging environment on AWS Beanstalk, the app will not start because we are getting  "Could not find [all of the development/test gems] in any of the sources" even though we have the env variable BUNDLE_WITHOUT set to "development:test"

deivid-rodriguez commented:
`BUNDLE_WITH` needs to be set both when installing and when starting the application. Is that the case?

gugaiz commented:
BUNDLE_WITHOUT is set as an environment variable on the OS so it is present on both (installing and starting)

deivid-rodriguez commented:
Well, then I don't know what's happening, sorry. If you give me instructions to reproduce this I can try to help.

gugaiz commented:
Trying to figure out what could be happening... so far I found that adding the following debug line into `bundler/definition.rb:482:in 'materialize'` function 

```
puts "BUNDLE_WITHOUT=#{ENV['BUNDLE_WITHOUT']}"
```
I can see that `BUNDLE_WITHOUT` variable is not set but all other env variables are set, ie if on bash I do
 
```
$ printenv|grep BUNDLE_
BUNDLE_WITHOUT=test:development
```
I also noticed that if I run `bundle info spring` I get the following result
```
$ bundle info spring
BUNDLE_WITHOUT=test:development
Could not find gem 'spring', because it's in the group 'development', configured to be ignored.
```
meaning that it is set in that case.

It seems to be an issue related to beanstalk environments. I will create a ticket there.

nsmmrs commented:
@gugaiz On Heroku, the build stage ENV is configured separately from the normal application ENV. AWS may have a similar setup.
